enterprise/internal/insights/query/compute_executor.go

```diff
@@ -0,0 +1,128 @@
+package query
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/inconshreveable/log15"
+
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/insights/compression"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/insights/query/querybuilder"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/insights/query/streaming"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+type ComputeExecutor struct {
+	justInTimeExecutor
+	computeSearch func(ctx context.Context, query string) ([]GroupedResults, error)
+}
+
+func NewComputeExecutor(postgres database.DB, clock func() time.Time) *ComputeExecutor {
+	executor := ComputeExecutor{
+		justInTimeExecutor: justInTimeExecutor{
+			db:        postgres,
+			repoStore: postgres.Repos(),
+			filter:    &compression.NoopFilter{},
+			clock:     clock,
+		},
+		computeSearch: streamTextExtraCompute,
+	}
+
+	return &executor
+}
+
+func streamTextExtraCompute(ctx context.Context, query string) ([]GroupedResults, error) {
+	decoder, streamResults := streaming.ComputeTextDecoder()
+	err := streaming.ComputeTextExtraStream(ctx, query, decoder)
+	if err != nil {
+		return nil, err
+	}
+	if len(streamResults.Errors) > 0 {
+		return nil, errors.Errorf("compute streaming search: errors: %v", streamResults.Errors)
+	}
+	if len(streamResults.Alerts) > 0 {
+		return nil, errors.Errorf("compute streaming search: alerts: %v", streamResults.Alerts)
+	}
+	return computeTabulationResultToGroupedResults(streamResults), nil
+}
+
+func (c *ComputeExecutor) Execute(ctx context.Context, query, groupBy string, repositories []string) ([]GeneratedTimeSeries, error) {
+	repoIds := make(map[string]api.RepoID)
+	for _, repository := range repositories {
+		repo, err := c.repoStore.GetByName(ctx, api.RepoName(repository))
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to fetch repository information for repository name: %s", repository)
+		}
+		repoIds[repository] = repo.ID
+	}
+	log15.Debug("Generated repoIds", "repoids", repoIds)
+
+	groupedValues := make(map[string]int)
+	for _, repository := range repositories {
+		modifiedQuery, err := querybuilder.ComputeInsightCommandQuery(query, querybuilder.MapType(strings.ToLower(groupBy)))
+		if err != nil {
+			return nil, errors.Wrap(err, "ComputeInsightCommandQuery")
+		}
+		modifiedQuery = querybuilder.SingleRepoQueryIndexed(modifiedQuery, repository)
```

Comment:
```suggestion
        modifiedQuery = querybuilder.SingleRepoQueryIndexed(modifiedQuery, repository)
		modifiedQuery, err := querybuilder.ComputeInsightCommandQuery(query, querybuilder.MapType(strings.ToLower(groupBy)))
		if err != nil {
			return nil, errors.Wrap(err, "ComputeInsightCommandQuery")
		}
		
```

This is safer to modify to a compute type afterwards. I think I'll make this type-safe in the code to make that clear.

Reply:
https://github.com/sourcegraph/sourcegraph/pull/38315